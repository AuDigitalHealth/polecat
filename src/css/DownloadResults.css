.download-results {
  cursor: pointer;
}

.download-results.disabled {
  opacity: 0.3;
  cursor: default;
}
