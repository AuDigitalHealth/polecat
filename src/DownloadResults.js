import React, { Component } from 'react'
import PropTypes from 'prop-types'
import FileSaver from 'file-saver'

import Icon from './Icon.js'
import { codingToSnomedCode, codingToSnomedDisplay } from './fhir/medication'

import './css/DownloadResults.css'

class DownloadResults extends Component {
  static propTypes = {
    results: PropTypes.arrayOf(
      PropTypes.shape({
        coding: PropTypes.arrayOf(
          PropTypes.shape({
            system: PropTypes.string,
            code: PropTypes.string,
            display: PropTypes.string,
          }),
        ),
        type: PropTypes.string.isRequired,
      }),
    ),
    shownGMs: PropTypes.array,
    loading: PropTypes.bool,
    onClick: PropTypes.func,
  }
  static defaultProps = {
    results: [],
    shownGMs: [],
  }

  constructor(props) {
    super(props)
    this.handleClick = this.handleClick.bind(this)
  }

  downloadResults(results, shownGMs) {
    const header = ['SCTID', 'Preferred term', 'AMT concept type']
    // Add columns for the SCTID and preferred term of each generalized medicine.
    for (const shownGM of shownGMs) {
      header.push(`${shownGM} SCTID`)
      header.push(`${shownGM} preferred term`)
    }
    header.push('View in browser')
    let tsv = `${header.join('\t')}\n`
    for (const result of results) {
      const code = codingToSnomedCode(result.coding),
        display = codingToSnomedDisplay(result.coding),
        type = result.type,
        browserLink = `${this.getBrowserUrl()}/Medication/${code}`,
        row = [code, display, type]
      // Add SCTID and preferred term for each generalized medicine.
      for (const shownGM of shownGMs) {
        const foundGM = result.generalizedMedicines.find(
          m => m.type === shownGM,
        )
        if (foundGM) {
          row.push(codingToSnomedCode(foundGM.coding))
          row.push(codingToSnomedDisplay(foundGM.coding))
        }
      }
      row.push(browserLink)
      tsv += `${row.join('\t')}\n`
    }
    const blob = new Blob([tsv], { type: 'text/tab-separated-values' })
    FileSaver.saveAs(blob, 'results.tsv')
  }

  getBrowserUrl() {
    const { protocol, host } = window.location
    return `${protocol}//${host}`
  }

  handleClick(event) {
    const { onClick } = this.props
    if (onClick) onClick(event)
  }

  componentWillReceiveProps(nextProps) {
    const { results, shownGMs } = nextProps
    if (results && results !== this.props.results)
      this.downloadResults(results, shownGMs)
  }

  render() {
    const { loading } = this.props

    return (
      <Icon
        className={loading ? 'download-results disabled' : 'download-results'}
        type="download"
        hoverType="download-active"
        width={21}
        height={21}
        alt="Download search results as TSV"
        title="Download search results as TSV"
        onClick={loading ? null : this.handleClick}
      />
    )
  }
}

export default DownloadResults
