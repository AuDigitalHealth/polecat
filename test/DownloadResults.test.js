import React from 'react'
import { shallow } from 'enzyme'
import sinon from 'sinon'
import FileSaver from 'file-saver'

import DownloadResults from '../src/DownloadResults.js'

describe('DownloadResults', () => {
  it('should report click to handler when clicked', () => {
    const onClick = jest.fn()
    const props = {
      onClick,
    }
    const wrapper = shallow(<DownloadResults {...props} />)
    wrapper.simulate('click')
    expect(onClick).toHaveBeenCalled()
  })

  it('should download results when they are passed through props', () => {
    const saveAs = sinon.stub(FileSaver, 'saveAs')
    const props = {
      results: [
        {
          type: 'TPUU',
          coding: [
            {
              system: 'http://snomed.info/sct',
              code: '725751000168107',
              display: 'PKU Cooler 20 Green oral liquid solution, 174 mL pouch',
            },
          ],
          status: 'active',
          sourceCodeSystemUri: 'http://snomed.info/sct',
          sourceCodeSystemVersion:
            'http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228',
        },
        {
          type: 'TPUU',
          coding: [
            {
              system: 'http://snomed.info/sct',
              code: '710331000168102',
              display:
                'PKU Air 20 Red (Berry Blast) oral liquid solution, 174 mL pouch',
            },
          ],
          status: 'active',
          sourceCodeSystemUri: 'http://snomed.info/sct',
          sourceCodeSystemVersion:
            'http://snomed.info/sct?version=http%3A%2F%2Fsnomed.info%2Fsct%2F32506021000036107%2Fversion%2F20180228',
        },
      ],
    }
    const wrapper = shallow(<DownloadResults />)
    wrapper.setProps(props)
    expect(saveAs.calledOnce).toBe(true)
    return new Promise(resolve => {
      const reader = new FileReader()
      reader.addEventListener('load', event => {
        expect(event.target.result).toMatchSnapshot()
        resolve()
      })
      reader.readAsText(saveAs.firstCall.args[0])
    })
  })

  it('should not register click when loading', () => {
    const onClick = jest.fn()
    const props = {
      loading: true,
      onClick,
    }
    const wrapper = shallow(<DownloadResults {...props} />)
    wrapper.simulate('click')
    expect(onClick).not.toHaveBeenCalled()
  })
})
