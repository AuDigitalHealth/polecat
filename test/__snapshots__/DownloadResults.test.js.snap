// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadResults should download results when they are passed through props 1`] = `
"SCTID	Preferred term	AMT concept type	View in browser
725751000168107	PKU Cooler 20 Green oral liquid solution, 174 mL pouch	TPUU	http://localhost/Medication/725751000168107
710331000168102	PKU Air 20 Red (Berry Blast) oral liquid solution, 174 mL pouch	TPUU	http://localhost/Medication/710331000168102
"
`;
